helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -385,10 +389,19 @@ public static int getStateCount(String state, StateModelDefinition stateModelDef
           // If the desired state is the top state, but the instance cannot be transited to the
           // top state in one hop, try to keep the top state on current host or a host with a closer
           // state.
-          for (String currentStatePrioritizedInstance : sortedPreferenceList) {
-            if (!assigned.contains(currentStatePrioritizedInstance) && liveAndEnabled
-                .contains(currentStatePrioritizedInstance)) {
+          while (currentStatePrioritizedInstanceIter.hasNext()) {
+            // Note that it is safe to check the prioritized instance items only once here.
+            // Since the only possible condition when we don't use an instance in this list is that
+            // it has been assigned with some state. And this is not revertable in this method. So
+            // checking it one more time later will only waste time.
+            String currentStatePrioritizedInstance = currentStatePrioritizedInstanceIter.next();
```

Comment:
Maybe I am wrong. But could that be a scenario of:
preference list = 1, 2, 3
current state = 4, 5, 6

Then we keep assign ONLINE to 4,5,6. There will not be any movement. For DelayedAutoRebalance we may dont have it. But for default AutoRebalancer Or SemiAuto, we may use this for AbstractRebalancer then there will not be any movement.

Reply:
currentStatePrioritizedInstance != current state instances.
In your case, the currentStatePrioritizedInstance will be 1,2,3.